### PR TITLE
Update Mean-Teacher and FixMatch Self-Training Scheme(s)

### PR DESCRIPTION
**Pull request by anwai98**
- Mean-Teacher training parameters updated
- AdaMT training parameters updated

### CODE REVIEW

File: experiments/probabilistic_domain_adaptation/livecell/common.py [hunk 1]
```diff
@@ -55,9 +55,9 @@ def get_unet():
     return UNet2d(in_channels=1, out_channels=1, initial_features=64, final_activation="Sigmoid", depth=4)
 
 
-def load_model(model, ckpt, state="model_state", device=None):
-    state = torch.load(os.path.join(ckpt, "best.pt"))[state]
-    model.load_state_dict(state)
+def load_model(model, ckpt, get_model=get_unet, device=None):
```
Comment (constantinpape) [inline experiments/probabilistic_domain_adaptation/livecell/common.py:58]: You can now use this function for loading the model, it takes care of the pytorch 2 things: https://github.com/constantinpape/torch-em/blob/main/torch_em/util/util.py#L204

Comment (anwai98) [reply experiments/probabilistic_domain_adaptation/livecell/common.py]: Updated in the recent [commit](https://github.com/constantinpape/torch-em/pull/116/commits/18f289e6d6d8067ed625d80be0107b01bf9ab7ac#diff-414c225623f833e81230091d9429fed78c30d5a5c3f6fd26238ff9587d59202cR126).

File: torch_em/self_training/fix_match.py [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+        for xu1, xu2 in self.unsupervised_train_loader:
+            xu1, xu2 = xu1.to(self.device), xu2.to(self.device)
+
+            teacher_input, model_input = xu1, xu2
+
+            self.optimizer.zero_grad()
+            # Perform unsupervised training
+            with forward_context():
+                # Compute the pseudo labels.
+                pseudo_labels, label_filter = self.pseudo_labeler(self.model, teacher_input)
```
Comment (constantinpape) [inline torch_em/self_training/fix_match.py:164]: We need to be careful here that the pseudo labels are not taken into account when computing gradients. So this should be wrapped into a `with torch.no_grad()` and we should `detach` the `pseudo_labels`. (This might be missing from the mean teacher trainer as well and so should be added there too).

File: torch_em/self_training/fix_match.py [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+            with forward_context():
+                # We pass the model, the input and the labels to the supervised loss function,
+                # so that how the loss is calculated stays flexible, e.g. to enable ELBO for PUNet.
+                supervised_loss = self.supervised_loss(self.model, xs, ys)
+
+            teacher_input, model_input = xu1, xu2
+            # Perform unsupervised training
+            with forward_context():
+                # Compute the pseudo labels.
+                pseudo_labels, label_filter = self.pseudo_labeler(self.model, teacher_input)
```
Comment (constantinpape) [inline torch_em/self_training/fix_match.py:210]: Again, we need to make sure that no gradients are accumulated in the pseudo label computation (see above).

**Review by constantinpape — COMMENTED** [review]
Thanks! I left a few comments. The most important thing to address is making sure that pseudo labels are not used in gradient computation.

File: torch_em/self_training/fix_match.py [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        self.pseudo_labeler = pseudo_labeler
+
+        if source_distribution is None:
+            self.source_distribution = None
+        else:
+            self.source_distribution = torch.FloatTensor(source_distribution).to(self.device)
+
+        self._kwargs = kwargs
+
+    def get_distribution_alignment(self, pseudo_labels, label_threshold=0.5):
```
Comment (constantinpape) [inline torch_em/self_training/fix_match.py:130]: I didn't fully go through this, and I assume you adapted it from our previous code for the distro alignment.
Still, it would be good to check this again and to give some comments / formulas here.

Comment (constantinpape) [inline torch_em/self_training/fix_match.py:130]: I didn't fully go through this, and I assume you adapted it from our previous code for the distro alignment.
Still, it would be good to check this again and to give some comments / formulas here.

File: torch_em/self_training/fix_match.py [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+    Parameters:
+        model [nn.Module] -
+        unsupervised_train_loader [torch.DataLoader] -
+        unsupervised_loss [callable] -
+        supervised_train_loader [torch.DataLoader] - (default: None)
+        supervised_loss [callable] - (default: None)
+        unsupervised_loss_and_metric [callable] - (default: None)
+        supervised_loss_and_metric [callable] - (default: None)
+        logger [TorchEmLogger] - (default: SelfTrainingTensorboardLogger)
+        momentum [float] - (default: 0.999)
```
[the remark this answers is not in the document]
Comment (anwai98) [reply torch_em/self_training/fix_match.py:58]: Added the description (for both FixMatch and MeanTeacher trainers)

File: torch_em/self_training/fix_match.py [hunk 6]
```diff
@@ -0,0 +1,309 @@
...
+        - Returns: loss, metric
+    At least one of unsupervised_loss_and_metric and supervised_loss_and_metric must be given.
+
+    Note: adjust the batch size ratio between the 'unsupervised_train_loader' and 'supervised_train_loader'
+    for setting the ratio between supervised and unsupervised training samples
+
+    Parameters:
+        model [nn.Module] -
+        unsupervised_train_loader [torch.DataLoader] -
+        unsupervised_loss [callable] -
```
[the remark this answers is not in the document]
Comment (anwai98) [reply torch_em/self_training/fix_match.py:51]: Added the description (for both FixMatch and MeanTeacher trainers)

File: torch_em/self_training/fix_match.py (hunk 2 above)
Comment (anwai98) [reply torch_em/self_training/fix_match.py]: Fixed it (in both FixMatch and MeanTeacher trainers). Thanks for mentioning!

File: torch_em/self_training/fix_match.py (hunk 4 above)
Comment (anwai98) [reply torch_em/self_training/fix_match.py:135]: Added some highlights about distribution alignment from [ReMixMatch](https://arxiv.org/abs/1911.09785) (giving hints about the idea).

**Comment by anwai98** [comment]
Thanks for all the feedback! The expected changes have been made (in both the self-training approaches).

File: torch_em/self_training/fix_match.py (hunk 3 above)
Comment (constantinpape) [reply torch_em/self_training/fix_match.py:221]: You need to keep the `forward_context()` here. And I think you can just write
```python
with forward_context(), torch.no_grad():
   ...
```

File: torch_em/self_training/mean_teacher.py [hunk 7]
```diff
@@ -196,11 +197,13 @@ def _train_epoch_unsupervised(self, progress, forward_context, backprop):
 
             teacher_input, model_input = xu1, xu2
 
+            with torch.no_grad():
```
Comment (constantinpape) [inline torch_em/self_training/mean_teacher.py:200]: You should also add the `forward_context()` to the `with` statement. (See also above).

File: torch_em/self_training/mean_teacher.py [hunk 8]
```diff
@@ -244,9 +247,13 @@ def _train_epoch_semisupervised(self, progress, forward_context, backprop):
                 supervised_loss = self.supervised_loss(self.model, xs, ys)
 
             teacher_input, model_input = xu1, xu2
+
+            with torch.no_grad():
```
Comment (constantinpape) [inline torch_em/self_training/mean_teacher.py:251]: `forward_context`, see above.

File: torch_em/self_training/fix_match.py [hunk 9]
```diff
@@ -157,14 +162,20 @@ def _train_epoch_unsupervised(self, progress, forward_context, backprop):
 
             teacher_input, model_input = xu1, xu2
 
+            with torch.no_grad():
```
Comment (constantinpape) [inline torch_em/self_training/fix_match.py:165]: We also need the `forward_context` here, see below.

**Review by constantinpape — COMMENTED** [review]
Looks good overall, but we need to keep the `forward_context` around the pseudo_labeler.

**Comment by anwai98** [comment]
The `forward_context` has been updated for generating the Pseudo Labels. Thanks!

**Review by constantinpape — APPROVED** [review]
Thanks, looks all good now!
